tests/gen/gen_auto.go

```diff
@@ -0,0 +1,251 @@
+	types        map[string]client.CollectionDefinition
+	config       configsMap
+	resultDocs   []GeneratedDoc
+	usageCounter typeUsageCounters
+	cols         map[string][]docRec
+	docsDemand   map[string]typeDemand
+	random       rand.Rand
+}
+
+func (g *randomDocGenerator) GenerateDocs(options ...Option) ([]GeneratedDoc, error) {
+	g.cols = make(map[string][]docRec)
+	if g.config == nil {
+		g.config = make(configsMap)
+	}
+
+	configurator := newDocGenConfigurator(g.types, g.config)
+	err := configurator.Configure(options...)
+	if err != nil {
+		return nil, err
+	}
+
+	g.random = *configurator.random
+
+	g.docsDemand = configurator.docsDemand
+	g.usageCounter = configurator.usageCounter
+
+	g.resultDocs = make([]GeneratedDoc, 0, g.getMaxTotalDemand())
+	docsLists, err := g.generateRandomDocs(configurator.typesOrder)
+	if err != nil {
+		return nil, err
+	}
+	for _, docsList := range docsLists {
+		typeDef := g.types[docsList.ColName]
+		for _, doc := range docsList.Docs {
+			g.resultDocs = append(g.resultDocs, GeneratedDoc{
+				ColName: typeDef.Description.Name,
+				JSON:    createDocJSON(&typeDef, doc),
+			})
+		}
+	}
+	return g.resultDocs, nil
+}
+
+func (g *randomDocGenerator) getMaxTotalDemand() int {
+	totalDemand := 0
+	for _, demand := range g.docsDemand {
+		totalDemand += demand.max
+	}
+	return totalDemand
+}
+
+// getNextPrimaryDocKey returns the key of the next primary document to be used as a relation.
+func (g *randomDocGenerator) getNextPrimaryDocKey(secondaryType string, field *client.FieldDescription) string {
+	ind := g.usageCounter.getNextTypeIndForField(secondaryType, field)
+	docKey := g.cols[field.Schema][ind].docKey
+	return docKey
+}
+
+func (g *randomDocGenerator) getDocKey(typeDef *client.CollectionDefinition, doc map[string]any) (string, error) {
+	clientDoc, err := client.NewDocFromJSON([]byte(createDocJSON(typeDef, doc)))
+	if err != nil {
+		return "", err
+	}
+	return clientDoc.Key().String(), nil
+}
+
+func (g *randomDocGenerator) generateRandomDocs(order []string) ([]DocsList, error) {
+	result := []DocsList{}
+	for _, typeName := range order {
+		col := DocsList{ColName: typeName}
+		typeDef := g.types[typeName]
+
+		currentTypeDemand := g.docsDemand[typeName]
+		averageDemand := currentTypeDemand.getAverage()
+		for i := 0; i < averageDemand; i++ {
+			newDoc := make(map[string]any)
+			for _, field := range typeDef.Schema.Fields {
+				if field.Name == request.KeyFieldName {
+					continue
+				}
+				if field.IsRelation() {
+					if field.IsPrimaryRelation() {
+						newDoc[field.Name+request.RelatedObjectID] = g.getNextPrimaryDocKey(typeName, &field)
+					}
+				} else {
+					fieldConf := g.config.ForField(typeName, field.Name)
+					newDoc[field.Name] = g.generateRandomValue(typeName, field.Kind, fieldConf)
+				}
+			}
+			docKey, err := g.getDocKey(&typeDef, newDoc)
+			if err != nil {
+				return nil, err
+			}
+			g.cols[typeName] = append(g.cols[typeName], docRec{doc: newDoc, docKey: docKey})
```

Comment:
suggestion: It feels like this would be easier to read if the population of `g.cols` with `typeName` was done outside of the block that creates documents.  This thought might however be related to my uncertainty as to why `order` is important in this function, I assume it is, and I assume it is to do with the order in which documents must be generated, so I also suggest documenting why type order is important to this function.